dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,663 @@
+   * never fails. Calling this method will send all requests. The promise
+   * resolves immediately if there are no pending writes.
+   *
+   * @return {Promise<void>} A promise that resolves when there are no more
+   * pending writes.
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   *
+   * bulkWriter.create(documentRef, {foo: 'bar'});
+   * bulkWriter.update(documentRef2, {foo: 'bar'});
+   * bulkWriter.delete(documentRef3);
+   * await close().then(() => {
+   *   console.log('Successfully executed all writes');
+   * });
+   */
+  close(): Promise<void> {
+    this.verifyNotClosed();
+    this.closed = true;
+    return this._flush();
+  }
+
+  private async _flush(): Promise<void> {
+    const trackedBatches = this.batchQueue;
+    const writePromises = trackedBatches.map(batch => batch.awaitBatch());
+    this.sendBatches();
+    await Promise.all(writePromises);
+  }
+
+  private verifyNotClosed(): void {
+    if (this.closed) {
+      throw new Error('BulkWriter has already been closed.');
+    }
+  }
+
+  /**
+   * Gradually ramps up the maximum number of operations per second by 50%
+   * every 5 minutes in line with the 500/50/5 rule.
+   */
+  private rampMaxOps(): void {
+    delayExecution(() => {
+      this.maxOpsPerSecond = this.maxOpsPerSecond * 1.5;
+      this.rampMaxOps();
+    }, 5 * 60 * 1000);
+  }
+
+  /**
+   * Return the first eligible batch that can hold a write to the provided
+   * reference, or creates one if no eligible batches are found.
+   */
+  private getEligibleBatch(ref: DocumentReference): BulkCommitBatch {
+    let batch = null;
+    let logWarning = this.batchQueue.length > 0;
+    for (const bulkBatch of this.batchQueue) {
+      if (!bulkBatch.refsInBatch.has(ref)) {
+        logWarning = false;
+      }
+      if (bulkBatch.canAddDoc(ref)) {
+        batch = bulkBatch;
+      }
+    }
+
+    if (batch === null) {
+      if (logWarning) {
+        console.warn(
+          '[BulkWriter]',
+          'Writing to the same document multiple times will slow down BulkWriter. ' +
+            'Write to unique documents in order to maximize throughput.'
+        );
+      }
+      return this.createNewBatch();
+    } else {
+      return batch;
+    }
+  }
+
+  private createNewBatch(): BulkCommitBatch {
+    const newBatch = new BulkCommitBatch(
+      this.getId(),
+      this.firestore.batch(),
+      this.maxBatchSize
+    );
+
+    this.batchQueue.push(newBatch);
+    return newBatch;
+  }
+
+  private getId(): number {
+    const id = this.currentId;
+    if (this.currentId === Number.MAX_SAFE_INTEGER) {
+      this.currentId = 0;
+    } else {
+      this.currentId++;
+    }
+    return id;
+  }
+
+  /**
+   * Attempts to send batches starting from the front of the BatchQueue until a
+   * batch cannot be sent.
+   *
+   * After a batch is complete, try sending batches again.
+   */
+  private sendBatches(): void {
+    if (this.batchQueue.length === 0) {
```

Comment:
Does this short-circuit "if" statement provide meaningful savings?  It looks like the algorithm below will gracefully handle an empty batchQueue, nullifying the benefits of this short circuit.  If that is the case, I'd recommend removing this "if" statement because its presence IMO incorrectly implies that it provides a meaningful performance benefit and therefore detracts from readability.